### ISSUE TITLE
Firefox Addon

### ISSUE THREAD

**Issue by zettelzottel** [uav4geo/LibreTranslate#55]
Would be nice if you could select text and translate with your libtranslate instance

**Comment by pierotofy** [comment]
Agreed! @zettelzottel we'd welcome a contribution to add such an addon.

**Comment by Extarys** [comment]
What would you prefer? An icon than open a small window under it to translate? Or a sidebar?

This would really be nice if LibreTranslate could translate whole webpages as the addon could detect the page language and offer to translate it :)

**Comment by zettelzottel** [comment]
> Agreed! @zettelzottel we'd welcome a contribution to add such an addon. Interested?

I don't think I would find the time to do that and I have no experience writing an addon for firefox. It would be interesting to do such a task, because I have other things that would be cool as an addon, but it would take me forever to complete this…

**Comment by jonatasbaldin** [comment]
There's [this Firefox extension](https://github.com/FilipePS/Traduzir-paginas-web) which translates pages using Google Translate or Yandex. Would be nice to extend it to use a LibreTranslate backend.

**Comment by PJ-Finlay** [comment]
Argos Translate supports [translating HTML](https://github.com/argosopentech/translate-html),  we'd need to add support to the LibreTranslate API.

https://community.libretranslate.com/t/how-can-i-use-translate-html/58

**Comment by pierotofy** [comment]
Yep. I've been meaning to find some time to add HTML support to LibreTranslate.

Hopefully within the next few weeks.

**Comment by pierotofy** [comment]
HTML support is in (thanks to @dingedi).

**Comment by hugopeixoto** [comment]
I was trying to build this extension, but passing the whole HTML was giving me a gateway timeout. I was able to work around the issue by passing the contents of each `<p>` to a separate call to the API, but it take a lot of time to translate a full page.

I was testing with https://hermannschule.de/hermannpost.html and it took ~10 minutes to go through every paragraph and translate it. Do you all see this level of performance on your instances? Anything I might be doing wrong? I'm running it with `podman run --rm -it -p 5000:5000 docker.io/libretranslate/libretranslate` on a dual core VPS and API calls peg both cores to 100%.

**Comment by hugopeixoto** [comment]
Thanks for the info! I must be doing something wrong then, because it's taking 10 seconds to translate 1 sentence instead. 

In case I'm not able to figure out why this is so slow and I stop working on this, here's a repository with what I have so far:
https://github.com/hugopeixoto/libretranslate-firefox

- the API domain needs to be changed (translate.js -> DOMAIN) before it can be used
- it only translates `<p>` texts for now
- it is using `de` as the source language since the `auto` setting was not working most of the time (`WARNING:polyglot.detect.base:Detector is not able to detect the language reliably.`)

I was thinking of adding a randomly generated DOM attribute to each paragraph/element when I fetch them so I can map them more accurately when setting the translated text (in case the page changes or something).

**Comment by lawl** [comment]
Hey, I was also interested in something like this and decided to try my hand. You can find it here: https://github.com/lawl/translate

Still very early and rough around the edges. I don't think translating the entire page is a good idea and my addon doesn't do that, instead it tries to translate only what's visible on screen. So when you scroll down currently, you'll see it noticing new stuff and translating it. All of this could still be way more polished by translating slightly further down than what's visible on the viewport so one doesn't notice it while scrolling.

But again, early proof of concept. You can see a quick demo video here:

https://user-images.githubusercontent.com/680451/157904703-e5b5118d-4570-45d1-9e27-3726d9f59d1a.mp4

 Side note, is there an official icon, and what license is it under? I'm currently using "LibreTranslate" in the manifest.json for the addon name, but I'll of course change it if you're unhappy about me blatantly stealing your brand.

**Comment by pierotofy** [comment]
That's awesome!!

We follow the same policy as Mozilla's in terms of usage for the name LibreTranslate and icon. https://www.mozilla.org/en-US/foundation/trademarks/policy/

If you're an official LibreTranslate project it's OK using the name LibreTranslate and icon for the add-on (would you be interested in making this add-on an official project? I think it would be a great fit)

If you're developing the add-on as an independent third party, a different name would be more appropriate as to avoid confusion for users.

**Comment by pierotofy** [comment]
Sure! Feel free to get in touch even in private if you want to chat. https://uav4geo.com/contact

**Comment by lawl** [comment]
So, the issue with this is addon is currently a very annoying situation where chrome requires Manifest v3 for addons at least if you want to publish it in the web-store, but Firefox only supports Manifest v2 as for now. Unfortunately the require quite different code architectures and I'm currently unwilling to maintain two independent addons.

So I've decided to remove the libretranslate name for now and only support manifest v3 and keep this addon mostly for personal use for now. I'll re-evaluate seriously maintaining a browser extension once it becomes clearer how manifest v3/v2 plays out and what all the chromium forks will do.

**Comment by Extarys** [comment]
I hope FF contributors will come to a concensus soon on what to do with V3 :/ Google is such an annoying company.

I love this line very much: https://github.com/lawl/translate/blob/master/popup.js#L95

**Comment by K-Francis-H** [comment]
Hi, I've also created a Firefox extension: [https://github.com/K-Francis-H/libretranslate-unofficial-ff-extension](https://github.com/K-Francis-H/libretranslate-unofficial-ff-extension) 

It features inline translation: highlight text with your cursor and hold the Alt key and it will translate the highlighted text, release the Alt key and it will return to its original content. The inline feature is very useful for vocabulary acquisition and checking your reading comprehension in a second language. It also has the standard textbox translation interface in the popup. It can also be configured to use a self-hosted LibreTranslate API or any public one on the web.

I haven't released it on the Add-on store yet because by default it uses the LibreTranslate instance at [https://translate.argosopentech.com](https://translate.argosopentech.com) and I don't have their permission to drive traffic there. I was either going to get someone's permission to use their instance by default, or spin up my own server.

**Comment by PJ-Finlay** [comment]
You're welcome to use [translate.argosopentech.com](https://translate.argosopentech.com) to get started but if you get a lot of users (hopefully you do!) please set up you own instance.

**Comment by K-Francis-H** [comment]
It looks like Mozilla approved the Addon. It's available here: [https://addons.mozilla.org/en-US/firefox/addon/open-libretranslate-addon/](https://addons.mozilla.org/en-US/firefox/addon/open-libretranslate-addon/) If anyone has feedback, leave a review on the Addon store or open an Issue or PR at the github repo: [https://github.com/K-Francis-H/libretranslate-unofficial-ff-extension](https://github.com/K-Francis-H/libretranslate-unofficial-ff-extension)

@PJ-Finlay Thanks for letting me use [https://translate.argosopentech.com/](https://translate.argosopentech.com/) . If I get ~100 users I'll setup a server and change the default URL. If anytime before that you are seeing too much traffic, or would just prefer I stop using your server for any reason, just reach out and let me know. I need to setup a personal website for some other things anyway, so I'll try to prioritize hosting my own LibreTranslate instance as my next side project.

**Comment by K-Francis-H** [comment]
I got my own LibreTranslate server setup at [translate.terraprint.co](https://translate.terraprint.co) . Setting that up and getting SSL in place was actually a lot easier than I thought it would be. So I was going to update my Addon in the next few days to use my instance as well as fix a few other things too.

I noticed a recent update to the LibreTranslate README added trademark guidelines. Since I was going to update the Addon anyway, is there anything about the [addon store listing](https://addons.mozilla.org/en-US/firefox/addon/open-libretranslate-addon/) that you would like me to change to avoid confusion that my addon is officially associated with LibreTranslate? The icon I hastily came up with has a superficial resemblance to LibreTranslate's icon and I ended up with the kinda awkward listing title "Open LibreTranslate Addon" because "Unofficial LibreTranslate Addon" is too long for Mozilla (31 character limit).  "Unofficial LibreTranslate Addon" is used in the manifest and popup UI though. I also make it clear that the addon is unofficial in the listing description. If anything is a concern as far as LibreTranslate's trademarks are concerned though, just let me know and I'll try to make the appropriate adjustments to the addon.

**Comment by pierotofy** [comment]
Thanks for checking! I think the listing if fine, perhaps just don't call it "Open LibreTranslate Addon" (it contains the name LibreTranslate), maybe just "Open Translate Addon" or something else.

If you want to use the name LibreTranslate we could discuss options for joining the project officially otherwise. :pray:

**Comment by vitonsky** [comment]
Hi all. I've search autonomic machine learning translator for my browser extension and i ended up here.

If you guys need a browser extension, try a Linguist: https://github.com/translate-tools/linguist
Now Linguist are support custom translators, more info about it here: https://github.com/translate-tools/linguist/blob/master/docs/CustomTranslator.md

In short, you can make JS wrapper to your any API and embedded your translator to Linguist in settings.

Example of translator which use LibreTranslate API:
```js
function LibreTranslator() {
    const translator = new BaseTranslator();
    translator.translate = (text, from, to) => {
        var form = new FormData();
        form.append("q", text);
        form.append("source", from);
        form.append("target", to);
        form.append("format", "text");
        form.append("api_key", "");

        return fetch("https://translate.terraprint.co/translate", {
            credentials: "omit",
            headers: {
                "User-Agent":
                    "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:99.0) Gecko/20100101 Firefox/99.0",
                Accept: "*/*",
                "Accept-Language": "en-US,en;q=0.5",
                "Sec-Fetch-Dest": "empty",
                "Sec-Fetch-Mode": "cors",
                "Sec-Fetch-Site": "same-origin",
            },
            body: form,
            method: "POST",
            mode: "cors",
        })
            .then((r) => r.json())
            .then(({ translatedText }) => translatedText);
    };
    translator.translateBatch = (texts, from, to) =>
        Promise.all(texts.map((text) => this.translate(text, from, to)));

    translator.getLengthLimit = () => 4000;
    translator.getRequestsTimeout = () => 300;
    translator.checkLimitExceeding = () => -10000;

    Object.assign(this, translator);
}

LibreTranslator.isSupportedAutoFrom = () => true;
LibreTranslator.getSupportedLanguages = () => ["en", "ru", "ja", "de"];

LibreTranslator;
```

Just insert this code in settings and use this translator as well. Or you can deploy LibreTranslate locally or on your own server and use your own server.

Linguist is very powerful tool to translate any text and web pages, you can accumulate favorite words to your dictionary and configure all flexible, and this is open source (welcome to contribute) so probably it's that you need

**Comment by CodeCubeNeo** [comment]
Any updates on this? I especially like the idea of using Linguist.

**Comment by KubaIT** [comment]
@K-Francis-H 
(I got my own LibreTranslate server setup at [translate.terraprint.co](https://translate.terraprint.co/). Setting that up and getting SSL.)

How did you get this setup so you can access it via your own URL and not a local host IP?

**Comment by K-Francis-H** [comment]
@KubaIT 
I used nginx on a linode instance with a wildcard cert for *.terraprint.co to redirect incoming requests for translate.terraprint.co to port 5000 where I have the docker image of LibreTranslate running. Note that you might need to make a firewall rule to prevent direct HTTP access to port 5000. I noticed that it could also be reached at terraprint.co:5000 without HTTPS so I'm trying to fix that now.

**Comment by KubaIT** [comment]
Awesome. Thank you mate.

Best,
Kuba

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
